pvlib/iam.py

```diff
@@ -940,3 +941,360 @@ def schlick_diffuse(surface_tilt):
+        optimize_result = minimize(residual_function, guess, method="powell",
+                                   bounds=bounds)
+
+    if not optimize_result.success:
+        try:
+            message = "Optimizer exited unsuccessfully:" \
+                      + optimize_result.message
+        except AttributeError:
+            message = "Optimizer exited unsuccessfully: \
+                       No message explaining the failure was returned. \
+                       If you would like to see this message, please \
+                       update your scipy version (try version 1.8.0 \
+                       or beyond)."
+        raise RuntimeError(message)
+
+    return optimize_result
+
+
+def _process_return(target_name, optimize_result):
+    if target_name == "ashrae":
+        target_params = {'b': optimize_result.x.item()}
+
+    elif target_name == "martin_ruiz":
+        target_params = {'a_r': optimize_result.x.item()}
+
+    elif target_name == "physical":
+        L, n = optimize_result.x
+        target_params = {'n': n, 'K': 4, 'L': L}
+
+    return target_params
+
+
+def convert(source_name, source_params, target_name, options=None, fix_n=True):
+    """
+    Given a source model and its parameters and a target model, finds
+    parameters for the target model that best fit the source model.
+
+    Parameters
+    ----------
+    source_name : str
+        Name of the source model. Must be ``'ashrae'``, ``'martin_ruiz'``, or
+        ``'physical'``.
+
+    source_params : dict
+        A dictionary of parameters for the source model.
+
+            If source model is ``'ashrae'``, the dictionary must contain
+            the key ``'b'``.
+
+            If source model is ``'martin_ruiz'``, the dictionary must
+            contain the key ``'a_r'``.
+
+            If source model is ``'physical'``, the dictionary must
+            contain the keys ``'n'``, ``'K'``, and ``'L'``.
+
+    target_name : str
+        Name of the target model. Must be ``'ashrae'``, ``'martin_ruiz'``, or
+        ``'physical'``.
+
+    options : dict, optional
+        A dictionary that allows passing a custom weight function and
+        arguments for the weight function.
+        Keys of `options` can be ``'weight_function'`` and
+        ``'weight_args'``.
+
+            weight_function : function
+                A function that outputs an array of weights to use
+                when computing residuals between models.
+
+                Requirements:
+                    1. Must have ``aoi`` as the first positional argument.
+                    2. Any other arguments must be keyword arguments,
+                       defined in ``weight_args``.
+                    3. Must return a float or an array-like object with the
+                       same shape as ``aoi``.
+
+            weight_args : dict, optional
+                A dictionary containing keyword arguments for the
+                weight function.
```

Comment:
What's the idea behind supporting `weight_args`?  It seems like a user can achieve the same thing by defining a wrapper around their weight function and handing that to pvlib.  If so, I'd say let's require `weight_function` to be a single-parameter function and get rid of `weight_args` to simplify the API here.

Reply:
Replaced `options` with a kwarg `weight` that looks for a function value.